tools/cli/src/cargo_manifest.rs

```diff
@@ -34,27 +31,67 @@ pub(crate) fn extract_sdk_version(path: &Path) -> Result<Version, ManifestError>
             CargoTomlError::Parse(e) => e.into(),
             CargoTomlError::Io(e) => e.into(),
             CargoTomlError::InheritedUnknownValue => {
-                CannotProcessManifest("inherited unknown value".to_string())
+                ManifestError::CannotProcessManifest("inherited unknown value".to_string())
             }
-            CargoTomlError::WorkspaceIntegrity(reason) => CannotProcessManifest(reason),
-            CargoTomlError::Other(reason) => CannotProcessManifest(reason.to_string()),
-            _ => CannotProcessManifest("Unknown".to_string()),
+            CargoTomlError::WorkspaceIntegrity(reason) => {
+                ManifestError::CannotProcessManifest(reason)
+            }
+            CargoTomlError::Other(reason) => {
+                ManifestError::CannotProcessManifest(reason.to_string())
+            }
+            _ => ManifestError::CannotProcessManifest("Unknown".to_string()),
         }
     })?;
 
-    let dependency = manifest
+    if !manifest.dependencies.contains_key(SKD_CRATE_NAME) {
+        return Err(ManifestError::NonMarineWasm);
+    }
+
+    let package = manifest
+        .package
+        .ok_or_else(|| ManifestError::NoPackageInManifest(format!("{}", path.display())))?;
+
+    // TODO: find & load lockfile once per `marine build` run
+    path.ancestors()
+        .find_map(|path| extract_sdk_version_from_lockfile(&package, &path.join(LOCKFILE_NAME)))
+        .ok_or_else(|| ManifestError::CannotFindLockfile(format!("{}", path.display())))
+}
+
+fn extract_sdk_version_from_lockfile(
+    target_package: &cargo_toml::Package,
+    path: &Path,
+) -> Option<semver::Version> {
+    log::debug!("Trying to load lockfile from {}", path.display());
+    let lockfile = cargo_lock::Lockfile::load(path).ok()?;
```

Comment:
Suggest returning `Result` from `extract_sdk_version_from_lockfile` function and moving `CannotFindLockfile` creating to this point, because otherwise errors of not found package are equal to `CannotFindLockfile` that could confuse users

Reply:
I reworked all Cargo.lock stuff, needs re-review